Python/SDL/pyStateDescriptor.cpp

```diff
@@ -0,0 +1,183 @@
+/* This file is part of HSPlasma.
+ *
+ * HSPlasma is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * HSPlasma is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with HSPlasma.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "pySDL.h"
+
+#include <SDL/plStateDescriptor.h>
+
+PY_PLASMA_DEALLOC(StateDescriptor)
+PY_PLASMA_EMPTY_INIT(StateDescriptor)
+PY_PLASMA_NEW(StateDescriptor, plStateDescriptor)
+
+PY_PLASMA_LENGTH_DECL(StateDescriptor)
+{
+    return self->fThis->getNumVars();
+}
+
+PY_PLASMA_SUBSCRIPT_DECL(StateDescriptor)
+{
+    if (pyPlasma_check<size_t>(key)) {
+        try {
+            return pyVarDescriptor_FromVarDescriptor(self->fThis->get(pyPlasma_get<size_t>(key)));
+        } catch (...) {
+            PyErr_SetNone(PyExc_IndexError);
+            return nullptr;
+        }
+    } else if (pyPlasma_check<ST::string>(key)) {
+        plVarDescriptor* desc = self->fThis->get(pyPlasma_get<ST::string>(key));
+        if (!desc) {
+            PyErr_SetNone(PyExc_KeyError);
+            return nullptr;
+        }
+        return pyVarDescriptor_FromVarDescriptor(desc);
+    }
+
+    PyErr_SetString(PyExc_TypeError, "StateDescriptor subscript should be an int or a string");
+    return nullptr;
+}
+
+PY_PLASMA_ASS_SUBSCRIPT_DECL(StateDescriptor)
+{
+    if (!pyVarDescriptor_Check(value)) {
+        PyErr_SetString(PyExc_TypeError, "StateDescriptor subscript assignment expects a plVarDescriptor");
+        return -1;
+    }
+
+    if (pyPlasma_check<size_t>(key)) {
+        try {
+            self->fThis->set(pyPlasma_get<size_t>(key), ((pyVarDescriptor*)value)->fThis);
+        } catch (...) {
+            PyErr_SetNone(PyExc_IndexError);
+            return -1;
+        }
+
+        ((pyVarDescriptor*)value)->fPyOwned = false;
+        return 0;
+    } else if (pyPlasma_check<ST::string>(key)) {
+        // Inserts the variable if it does not exist already
+        ((pyVarDescriptor*)value)->fPyOwned = false;
+        self->fThis->set(pyPlasma_get<ST::string>(key), ((pyVarDescriptor*)value)->fThis);
+        return 0;
+    }
+
+    PyErr_SetString(PyExc_TypeError, "StateDescriptor subscript should be an int or a string");
+    return -1;
+}
+
+PY_PLASMA_REPR_DECL(StateDescriptor)
+{
+    return PyString_FromSTString(self->fThis->getName());
```

Comment:
The `repr` string should include more than just the name of the descriptor...  Maybe something like `<plStateDescriptor {name}>`?

Reply:
This is actually the implementation of the `__str__()` method, which shares its signature with `__repr__()`. I can introduce a proper `__repr__()` if you prefer.

Reply:
Sneaky; I didn't catch that you were assigning it to `tp_str` :trollface:.  In that case, I might prefer a new macro (even if it's almost identical to `PY_PLASMA_REPR_DECL`) to make it clearer in the code.